### DISCUSSION TITLE
Inherent prototype limitations and poor candidates for its design.

### DISCUSSION THREAD

**Discussion by kovitikus** [evennia/evennia#2531] [General Discussion]
In my prototypes, I'm setting various keys, that represent the key (name) of an object.

Take my torch object for example.

```py
CRUDELY_MADE_TORCH = {
    'prototype_parent': 'lighting',
    'prototype_key': 'crudely-made torch',
    'key': 'a crudely-made torch',
    'plural_key': 'crudely-made torches',
    'singular_key': 'a crudely-made torch',
    'typeclass': 'items.objects.Torch',
    'price': {'plat': 0, 'gold': 0, 'silver': 0, 'copper': 10},
    'fuel': 90,
    'burn_rate': 30
}
```

In this prototype, I have the normal key, the plural key, and the singular key. These two additional keys are important, because I use them when grouping objects or in other systems where I may have to reference a single torch or many torches. It's easier to set these here than it is to dynamically determine them when they are needed. Some words have peculiar rules about pluralization, so it's not viable to write logic that dynamically determines any given word's pluralization.

Now, I have another prototype where I spawn an inventory container, such as a bag.

```py
INVENTORY_CONTAINER = {
    'prototype_key': 'inventory_container',
    'key': 'inventory_container',
    'typeclass': 'items.objects.Object',
    'tags': [('inventory_container', 'container', None)],
    'attrs': [('max_slots', 50, None, None)]
}

INVENTORY_BAG = {
    'prototype_parent': 'inventory_container',
    'key': lambda: generate_random_bag_key(),
    'home': lambda: trash_bin()
}
```

As you can see, this inventory bag prototype calls on a function to dynamically generate the key for the bag. Let's take a look.

```py
def generate_random_bag_key():
    color = ('red', 'blue', 'green', 'yellow', 'black')
    adjective = ('tattered', 'worn', 'pristine', 'well-crafted', 'frayed')
    # a tattered red bag
    # a pristine yellow bag
    bag_key = f"a {random.choice(adjective)} {random.choice(color)} bag"
    return bag_key
```

You can see that there are various possible keys that can be generated with this function.

Here's the problem. I can't track which key was generated and assign the proper plural and singular keys to contain the same key, but with a slight variation.

The plural and singular keys can't call on this function again, because it will return a different random result.

I can't store the result of the bag's key function call anywhere in this module either, because prototypes are stored in dictionaries at the module level and so there is only one instance of this module in the server's memory. If two people were to try and generate a bag at the same time, any variable I store at the module level is shared as the same variable and could potentially be changed by other calls to the function at any time.

So then where does this leave me? It means that a prototype is NOT a useful method of managing this type of item.

Let's say I were to move the key generation logic to the `at_object_creation()` method on the typeclass.

```py
class Container(Object):
    pass

class InventoryContainer(Container):
    def at_object_creation(self):
        self.tags.add('inventory_container', category='container')
        self.attributes.add('max_slots', 50)
```

I could easily add my key generation method here, and store its result in a variable, and use that variable to assign the various key fields I require.

The problem is that the bag is only one type of possible inventory container. A player may want a satchel or a sack or a backpack, etc.

This means that I have to create a new typeclass for every one of these object types, because `at_object_creation()` can't be passed any requests to identify the object as a certain thing, as far as I'm aware. So instead I'd have to create each item as its own typeclass, to make sure that the proper key generation logic is used.

```py
class Container(Object):
    pass

class InventoryContainer(Container):
    def at_object_creation(self):
        self.tags.add('inventory_container', category='container')
        self.attributes.add('max_slots', 50)

class InventoryBag(InventoryContainer):
    def at_object_creation(self):
        # key generation logic here

class InventorySack(InventoryContainer):
    def at_object_creation(self):
        # key generation logic here

class InventorySatchel(InventoryContainer):
    def at_object_creation(self):
        # key generation logic here

class InventoryBackpack(InventoryContainer):
    def at_object_creation(self):
        # key generation logic here
```

This is a viable solution, but the entire point of the prototype system is to AVOID this sort of logic. In this situation I am hardcoding typeclasses into the game for some very minor information.

***

A possible solution? Handlers. Specifically my ItemHandler. Within it I could put my own `spawn()` method and handle all logic in there. This way I can pass in my request for a bag or a satchel, etc, while still being able to set the typeclass as Object (handling identification of the inventory container type with tags) and keeping all logic for the object's creation in a central location. 

```py
inv_container = character.item.spawn('inventory_bag')
character.equip.add(inv_container, slot='inventory_container')
```

It also means that I can dynamically generate the key I need for the object, and store that on a variable. This variable would be specific to whatever object is calling the ItemHandler, because handlers are instantiated objects themselves and therefore unique to the calling object.

I'd still have to write out all of the logic for handling things. When the `spawn()` method is passed an item name, I'll still have to deal with looking for that item's entry and then spawning it by using `create_object()` and manually applying all of my extra bits (`plural_key` and `singular_key`).

***

The prototype system is still useful for basic things that require no extra logic to call upon; logic that needs to be stored and referenced multiple times during object creation. Hell, it's even still useful for one-time logic calls (if I didn't require setting multiple names for my bag, for example).

But, the truth is that at the end of the day, if your items, NPCs, rooms, etc, have enough "intelligence" in their creation, there's a really good chance that prototypes will fall remarkably short of your needs.

**Comment by davewiththenicehat** [comment]
It has been a good long while since I have worked with prototypes. I remember each key in the prototype dictionary getting a dB entry on the object. Can you use those keys at_object_creation for the reference you need?

**Reply by kovitikus** [reply]
As far as I recall, the last time I spoke to Griatch about this, the prototype entries actually overwrite `at_object_creation`. Meaning that the `create_object()` method gets called first, probably, and then that triggers the `at_object_creation()` hook, which then returns the object to the prototype spawner, which then applies its properties to the object.

*Investigates.*

Ok, well the `spawn()` method calls `batch_create_object()` in the `spawner `module, which creates an object from `ObjectDB` from the `objects.models` module.

https://github.com/evennia/evennia/blob/master/evennia/prototypes/spawner.py#L769

Anyways, suffice to say it looks like `batch_create_object()` applies the prototype properties to the object after the object is generated, which I can only assume means that the `at_object_creation()` hook is called first.

And as I mentioned, I'm pretty sure Griatch confirmed that same thing to me awhile ago.

**Comment by davewiththenicehat** [comment]
I see. The call order logic. I will admit I do not understand the overall goal. I think it is determining if this object is multiple objects after object creation, when the object also has random keys.

Could you have your `generate_random_bag_key` do the logic you want after keys are assigned? These callbacks are called after object creation in `batch_creation_object`.

Also I notice `batch_create_object` returns the object(s) created. Does spawn method return them also? Could you do the tasks on the returned object(s)?

Would a function `obj.at_spawn_completion` resolve your issue? Where it is a function on the object that is called at the end of `batch_create_object`.
Much the way `obj.save` is automatically called at the end of `batch_create_object`.
https://github.com/evennia/evennia/blob/77c16a4779497a50ce34e675dd604a266bb5fe79/evennia/prototypes/spawner.py#L827

**Reply by kovitikus** [reply]
Yes, the `spawn()` method returns a list of 1 or more objects. I just don't want to do my logic in multiple places (every place I spawn the said item) in my code. Hence why I think that having a `spawn()` method on my ItemHandler would be my best approach, because I can define the object in my own dictionary somewhere (for the tags and stuff that are always the same) and then generate the random key for the item in my custom `spawn()` method.

The attributes I'm attempting to assign are preemptive strings, designed to allow me to easily change keys when grouping and ungrouping objects.

**Comment by strikaco** [comment]
> You can see that there are various possible keys that can be generated with this function.
> Here's the problem. I can't track which key was generated and assign the proper plural and singular keys to contain the same key, but with a slight variation.
> The plural and singular keys can't call on this function again, because it will return a different random result.

I don't think that's the only problem here. What if I don't want a shit-quality bag? I'd have to re-roll the whole item. You aren't allowing room for customization, which means you need to create prototypes for every single item, for every possible permutation of adjectives applicable to the thing. The position you're in right now seems like an inevitability.

Consider that `'plural_key': 'crudely-made torches'` is just duplicating effort-- literally _anything_ can be "crudely-made;" this isn't a property specific to torches that warrants an explicit torch definition just to change the name. All torches are torches. Some of them have less durability than others. Do those really need their own definition, or is durability something that could have been tweaked at time of instantiation?

Dynamic keys were a pain in the ass when I attempted it too. What ended up working for me is:

* Setting the most-generic version of the name as the key itself (`"bag"` instead of `"brown paper bag"`). Just do it-- resist the urge to complicate this.
* Track the adjectives applicable to the object separately-- `obj.db.adjectives = set()` (or add them to tags with category `adjectives` if you want adjectives to be cumulative).
* At init or during object creation, add the adjectives to the alias handler (optionally, using category `adjectives`).
* Use `get_display_name` to grab a random adjective from the pool and concatenate it with the key. Present that for display. You could even add this product back into the alias handler as well. 

It won't present as `small brown bag`, but `small bag` and `brown bag` will both be valid names for it by default. It's less nonsense for users to have to type as well-- if you do this right, `get brown` will work the same as `get bag` or `get brown bag`. I've been called a lot of things in my life, but nobody has ever tried to define my "key" as `"skinny white middle-aged incoherent male human"`.

> But, the truth is that at the end of the day, if your items, NPCs, rooms, etc, have enough "intelligence" in their creation, there's a really good chance that prototypes will fall remarkably short of your needs.

Scribblenauts uses something akin to prototypes. Every object you see is composed of a flattened stack of properties (`attrs`) inherited from a collection of adjectives (`prototype_parents`). This system has its limitations, but the pattern itself does work well enough to spawn literally anything you can think of, on handheld hardware from a decade ago, while also inviting continued commercial success.

I'd argue that you're applying _too much_ intelligence to object creation. You're overthinking things. Everything in existence is not thought-out or artisanally crafted; most things are mass-produced from some sort of template, sometimes by accident. That we are all unique is the product of entropy/manufacturing flaws, not explicit definition.

**Reply by kovitikus** [reply]
You are completely missing the point. 

1. This post is about the inherent flaw in not being able to store information between property and attribute assignment in a prototype, as well as the work flow of `at_object_creation()` being called first. So even if I decided the key for my bag in `at_object_creation()`, it would be overwritten by the prototype defined key. If I don't even assign a key in the prototype, or anything else of value, the prototype loses all worth.

2. `crudely-made torch` is just for testing. I haven't fully developed my system for object quality and how the adjective gets assigned. 

3. That's not how you are supposed to use tags. @Griatch has been complaining about people misusing them like this for awhile.

4. I will name my bags `a dark red well-tailored silk bag with intricate golden stitching` if I want. That's the way I like it and that's the way I'll name them.

**Reply by strikaco** [reply]
You are trying to swim upstream and blaming the river for flowing backwards.

None of this indicates problems with prototyping. All of your logic could be handled in `get_display_name`.

What possible benefit do you anticipate from packing every adjective into the key itself?

**Reply by kovitikus** [reply]
The key assignment and adjectives are just examples, they don't matter. The point is that you cannot reference other prototype fields in your logic. They live in isolation from one another.

**Reply by strikaco** [reply]
That got me too-- I once tried to implement something where by setting an attribute, related tags/attributes would also be added. I found no way to easily reference the parent object's keys.

As-is with prototypes as dicts that experiment never got far, but I'm having a little more luck converting prototypes to Dataclasses/TypedDicts and adding custom logic there. It is tedious and often feels like I'm creating a custom DSL/ORM though. Do not recommend at this time :)

**Reply by kovitikus** [reply]
I was thinking of building something more like this:

https://github.com/kovitikus/hecate/blob/master/rooms/instance_handler.py#L148

Which references a dictionary of dictionaries that I can set properties on:

https://github.com/kovitikus/hecate/blob/master/rooms/zones.py#L25

Handling the object creation myself allows me to design my own logic about how the final product is built.

Maybe I'm just derpy, but I wonder if something like a `before_object_creation` hook makes sense? Not that I think it would help with prototypes anyways. I like writing my own handler logic anyhow. I enjoy having full control over the entire thing.

Prototypes really are just a glorified `create_object`, but with some hierarchical structure attached. It would be easy to replicate prototype parents in my own logic.

**Reply by strikaco** [reply]
I think you and I are trying to achieve similar ends independently and through wildly different methods. Some clever ideas in your instance handler there.

Maybe you're onto something; I'm considering dropping the mixin/parent pattern altogether and dynamically composing objects from enum-based flags/adjectives. Add "enum:wings" to the stack at any time and `can_fly: True` and other properties are added to the object's attributes. Remove enum wings and these properties are undone, but the object might still be able to fly if it has enum jetpack or something.

Curious to see where our respective paths take us :)

**Comment by Griatch** [comment]
The issue is that each of your randomizing fields are independent from one another and as you point out there is really no way for one to know what result the other one produced when random. While there is a `$protkey()` protfunc in `develop` branch (for referencing other fields of the same prototype), this will still not work for dynamic content like your `lambda: generate_random_bag_key`. This is because even if the protfunc would execute it, you'd still get a different result for every call which would not help for your other pluralization fields etc. 

Prototype-values override the values set in `at_object_creation` because they are literally kwargs for the `create_object` function - and these will always take precedence over what is hard-coded in the hooks. This is as it should be.

One could speculate on a way to make this work: One would need to initialize the prototype to some _intermediary state_, where all executables have run - and _then_ apply protfuncs like  `$protkey` on it to get the one random value to apply to the others. So something like this: 

```
{
    randval = lambda: random.randint(1,10),
    key = "$get_key("$protkey(randval))"),
    plural = "$get_plural("$protkey(randval))") 
}
```
(this does _not_ work today, because that's not how prototypes work). That is, `randval` would get one random value, and then one would provide custom `$get_key/plural` protfuncs that uses `$protkey` to read the `randval` to get a single random value for all fields. It's an interesting thought, but an execution order like that is probably something for a PR and not something I'll spend time on any time soon.

 The obvious way to handle this right now would be to randomize the key in the prototype and then use that to determine your plural form at the typeclass level.

**Reply by kovitikus** [reply]
Thanks for the heads-up. I think I'll just use my ItemHandler for object creation that requires multiple steps that reference the same data. It's much easier to manage custom logic when I have full control over the object's creation process.

**Reply by davewiththenicehat** [reply]
Just for a reference. I know I had sent you a bit of info yesterday. An example of using the get_display_name is https://github.com/davewiththenicehat/uniqueMudSource/blob/c45a811031f2270d3203e5e1a8fbf643df55bb31/utils/emote.py#L60
With the um_emote function in that link, it allows you to replace /me and /target with pronouns.
I added support for sdesc. If I recall I even allows for aliases when the receiver has a customer alias. Where each player can recive the custom name have for a specific object.
It may be worth a look.
It is first person so it is really only good for your and my project. Most games are third person.

**Reply by kovitikus** [reply]
Yeah thanks. `get_display_name` is a great suggestion by @strikaco , I was just arguing that the particular hangup I'm having isn't specific to my adjective issue. I'll definitely take a look into refactoring how I present and set an object's key.
